### ISSUE TITLE
Slower imports after Rails 5.0 Upgrade

### ISSUE THREAD

**Issue by ranchodeluxe** [zdennis/activerecord-import#357]
After upgrading our application on Rails 4.1 ( with `activerecord-import` (0.15.0) ) to Rails 5.0.0.1 ( with `activerecord-import` (0.16.0) ) we noticed that the `import` times increased across different application environments and operating systems. All applications are using some minor version of MariaDB 5. I'd love some ideas about what I can do to help identify the issue.

To help isolate where the difference is I created a simple rake task. It selects existing records from the database, builds up an in memory array of values and imports them again into the table. 

```ruby
require 'ruby-prof'
namespace :bs do
  desc "running an import with 1000 features"
  task 'run:test' => :environment do |task, args|
    values = []
    Feature.first(1000).each{ |row|
      ctg_id = row.contiguous_id
      from = row.from
      to   = row.to
      comp = row.complement
      start_continues = row.start_continues
      stop_continues  = row.stop_continues
      values << ['fake', ctg_id, from, to, comp, start_continues, stop_continues, 'fake', ParentFeature.last.id, true]
    }
    options = { :validate => false }
    result = RubyProf.profile do
      Feature.import [:name, :contiguous_id, :from, :to, :complement, :start_continues, :stop_continues, :category, :parent_feature_id, :delta], values, options
    end

    printer = RubyProf::GraphHtmlPrinter.new(result)
    printer.print(File.open("/tmp/benchmark-graph-#{Time.now.strftime('%d-%m-%Y%H:%M')}.html",'w'), {})

    printer = RubyProf::CallStackPrinter.new(result)
    printer.print(File.open("/tmp/benchmark-callstack-#{Time.now.strftime('%d-%m-%Y%H:%M')}.html",'w'),{})
  end
end
```

I profiled it with `ruby-prof` ( the graph and callstack outputs can be seen in full [here](https://drive.google.com/drive/folders/0BwhqJE5CZmAgY0VSR3psYVlHVms?usp=sharing) ). The left-hand picture is `4.1` and the right-hand pictures is `5.0`. I'm wondering why the frame highlighted in red for `5.0` has so many more total calls that `4.1`?
![r41_versus_r5_activerecordimport](https://cloud.githubusercontent.com/assets/208859/21687044/86a56c44-d31c-11e6-9701-f8d293c6dc78.png)

Should I be using the `activerecord-importer` benchmarking tools to further investigate this? Any ideas are welcome. Thanks

**Comment by jkowens** [comment]
Can you go back to using activerecord-import 0.15.0 with Rails 5 and see if the import times decrease? That will just confirm that something in the 0.16.0 release is responsible.

**Comment by ranchodeluxe** [comment]
@jkowens: yeah, tried `v0.14.1` adn `v0.15.0` with Rails 5.0 and just as slow as `v0.16.0`

**Comment by jkowens** [comment]
Ok that's interesting. The only code that I can think of in activerecord-import that is specific to Rails 5 is here:

https://github.com/zdennis/activerecord-import/blob/master/lib/activerecord-import/import.rb#L673

**Comment by jkowens** [comment]
@ranchodeluxe it looks like type casting for Rails 5 is adding more overhead. I'm not sure if there is another way to go at it that performs better.

If anyone has any thoughts, please share!

**Comment by jkowens** [comment]
@ranchodeluxe I found that the type_caster being used needed to be memoized. Can you try installing activerecord-import from my branch for #359  and report back your results?

**Comment by ranchodeluxe** [comment]
@jkowens: nice work! yep, it's definitely faster. It's still not as fast as Rails 4.1. more about that below.
![type_cast_memo_callstack](https://cloud.githubusercontent.com/assets/208859/21745951/396a70c0-d4eb-11e6-9a14-d239253951e2.png)

I noticed quite a few extraneous calls in the `ruby-prof` graph related to type casting and quoting in Rails 5. I'm not saying those extraneous calls are responsible for the bottleneck. But it got me interested. 

For example, in my original post on this thread I'm asking why the number of `Array.map` calls in Rails 5 have increased so much in this workflow. The answer to that question is that the [datetime quoting workflow](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/mysql/quoting.rb#L31-L37) seems to check `ActiveRecord::ConnectionAdapters::AbstractAdapter::Version` for each date instead of once for all dates.
![typecast](https://cloud.githubusercontent.com/assets/208859/21745902/2d232dd0-d4ea-11e6-8951-2bf5c961d330.png)
 
Given that your PR also points to slow type casting, I decided to see how fast my simple test would be if I skipped type casting altogether ( since my simple test has the advantage of not needing to be cast ). I added an `:ignore_type_casting` option to `values_sql_for_columns_and_attributes`. Now it's blazing fast. Too bad we can't ignore the type casting for simpler situations. I think the question now becomes why is Rails 5 type casting so slow.

```ruby
    # Returns SQL the VALUES for an INSERT statement given the passed in +columns+
    # and +array_of_attributes+.
    def values_sql_for_columns_and_attributes(columns, array_of_attributes, ignore_type_casting) # :nodoc:
      # connection and type_caster get called a *lot* in this high intensity loop.
      # Reuse the same ones w/in the loop, otherwise they would keep being re-retreived (= lots of time for large imports)
      connection_memo = connection
      type_caster_memo = type_caster if respond_to?(:type_caster)

      array_of_attributes.map do |arr|
        my_values = arr.each_with_index.map do |val, j|
          column = columns[j]

          # be sure to query sequence_name *last*, only if cheaper tests fail, because it's costly
          if val.nil? && column.name == primary_key && !sequence_name.blank?
            connection_memo.next_value_for_sequence(sequence_name)
          elsif column
            if defined?(type_caster_memo) && type_caster_memo.respond_to?(:type_cast_for_database) # Rails 5.0 and higher
              if ignore_type_casting
                connection_memo.quote(val)
              else
                connection_memo.quote(type_caster_memo.type_cast_for_database(column.name, val))
              end
            elsif column.respond_to?(:type_cast_from_user)                                   # Rails 4.2 and higher
              connection_memo.quote(column.type_cast_from_user(val), column)
            else                                                                             # Rails 3.2, 4.0 and 4.1
              if serialized_attributes.include?(column.name)
                val = serialized_attributes[column.name].dump(val)
              end
              connection_memo.quote(column.type_cast(val), column)
            end
          end
        end
        "(#{my_values.join(',')})"
      end
    end
```

**Comment by ranchodeluxe** [comment]
After reviewing the `ruby-prof` callstack and graphs again, most of the type casting serializing time ( for my test at least ) seems to be related to working with automatic timestamps such as `updated_at` and `created_at` ( see the red section of the callstack below )

![timevalue_slog](https://cloud.githubusercontent.com/assets/208859/21747632/59192998-d522-11e6-97d4-426bba1667cb.png)

So we can get another speed increase if we disable automatic timestamp finagling with `ActiveRecord::Base.record_timestamps = false`

![disable_record_timestamps](https://cloud.githubusercontent.com/assets/208859/21747642/9dd6caea-d522-11e6-8233-fa8c4dc01406.png)

Here's the test import code now with the disabling of `record_timestamps`:
```ruby
require 'ruby-prof'
namespace :bs do
  desc "running an import with 1000 features"
  task 'run:test' => :environment do |task, args|
    values = []
    Feature.first(1000).each{ |row|
      ctg_id = row.contiguous_id
      from = row.from
      to   = row.to
      comp = row.complement
      start_continues = row.start_continues
      stop_continues  = row.stop_continues
      values << ['fake', ctg_id, from, to, comp, start_continues, stop_continues, 'fake', ParentFeature.last.id, true]
    }
    options = { :validate => false }

   ActiveRecord::Base.record_timestamps = false
   begin
      result = RubyProf.profile do
          Feature.import [:name, :contiguous_id, :from, :to, :complement, :start_continues, :stop_continues, :category, :parent_feature_id, :delta], values, options
      end
   ensure
      ActiveRecord::Base.record_timestamps = true # reenable it
   end

    printer = RubyProf::GraphHtmlPrinter.new(result)
    printer.print(File.open("/tmp/benchmark-graph-#{Time.now.strftime('%d-%m-%Y%H:%M')}.html",'w'), {})

    printer = RubyProf::CallStackPrinter.new(result)
    printer.print(File.open("/tmp/benchmark-callstack-#{Time.now.strftime('%d-%m-%Y%H:%M')}.html",'w'),{})
  end
end
```

**Comment by jkowens** [comment]
@ranchodeluxe nice detective work! I do see that ar-import offers a `:timestamps` option to disable adding the timestamps columns.

I wonder if the performance issue with dates is something that could be improved in activerecord? I see it's a call to `ActiveRecord::ConnectionAdapters::AbstractMysqlAdapter#supports_datetime_with_precision?` that leads to all of the version checks. Briefly looking at that code leads me to think memoizing an instance variable would eliminate redundant checks there.

**Comment by jkowens** [comment]
@ranchodeluxe I've released v0.17.0 that includes the update to memoize type_caster. I'm not sure what else can be done on this end. What do you think about opening an issue for ActiveRecord?

**Comment by jkowens** [comment]
I'm going to close this for now as it seems like this will have to be taken up in the Rails project.
